src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+    }
+
+    /// Clear transfer complete interrupt (tcif) for the DMA stream.
+    #[inline]
+    pub fn clear_transfer_complete_interrupt(&mut self) {
+        self.stream.clear_transfer_complete_interrupt();
+    }
+
+    /// Clear half transfer interrupt (htif) for the DMA stream.
+    #[inline]
+    pub fn clear_half_transfer_interrupt(&mut self) {
+        self.stream.clear_half_transfer_interrupt();
+    }
+
+    /// Clear transfer error interrupt (teif) for the DMA stream.
+    #[inline]
+    pub fn clear_transfer_error_interrupt(&mut self) {
+        self.stream.clear_transfer_error_interrupt();
+    }
+
+    /// Clear direct mode error interrupt (dmeif) for the DMA stream.
+    #[inline]
+    pub fn clear_direct_mode_error_interrupt(&mut self) {
+        self.stream.clear_direct_mode_error_interrupt();
+    }
+
+    /// Clear fifo error interrupt (feif) for the DMA stream.
+    #[inline]
+    pub fn clear_fifo_error_interrupt(&mut self) {
+        self.stream.clear_fifo_error_interrupt();
+    }
+
+    /// Changes the buffer and restarts or continues a double buffer transfer. This must be called
+    /// immediately after a transfer complete event. The closure must return `(BUF, T)` where `BUF`
+    /// is the new buffer to be used. The closure will not be called if the transfer is not
+    /// completed.
+    ///
+    /// # Panics
+    /// This method will panic when double buffering and one or both of the following conditions
+    /// happen:
+    ///
+    /// * The new buffer's length is smaller than the one used in the `init` method.
+    /// * The closure `f` takes too long to return and a buffer overrun occurs.
+    ///
+    /// # Safety
+    /// Memory corruption might occur if the previous buffer, the one passed to the closure, gets
+    /// dropped in the closure and an overrun occurs in double buffering mode.
+    pub unsafe fn next_transfer_with<F, T>(&mut self, f: F) -> Result<T, DMAError<()>>
+    where
+        F: FnOnce(BUF, CurrentBuffer) -> (BUF, T),
+    {
+        if !STREAM::get_transfer_complete_flag() {
+            return Err(DMAError::NotReady(()));
+        }
+        self.stream.clear_transfer_complete_interrupt();
+
+        if self.double_buf.is_some() && DIR::direction() != DmaDirection::MemoryToMemory {
+            let current_buffer = STREAM::current_buffer();
+            // double buffering, unwrap can never fail
+            let db = if current_buffer == CurrentBuffer::DoubleBuffer {
+                self.buf.take().unwrap()
+            } else {
+                self.double_buf.take().unwrap()
+            };
+            let r = f(db, !current_buffer);
+            let mut new_buf = r.0;
+            let (new_buf_ptr, new_buf_len) = new_buf.write_buffer();
+
+            // We can't change the transfer length while double buffering
+            assert!(
+                new_buf_len >= usize::from(self.transfer_length),
+                "Second Buffer not big enough"
+            );
```

Comment:
In this case we already gave the old buffer to the user through the closure, so we can't take it back, and we don't have anything to put in its place, so panic seems okay, we could also shutdown everything and return an error, but this seems more like a code bug instead of runtime error type thing, which would be quite inconvenient for the user to recover anyways.

Reply:
I agree, didn't think about how to continue the transfer in this case.